packages/dds/sequence/src/revertibles.ts

```diff
@@ -0,0 +1,185 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { unreachableCase } from "@fluidframework/common-utils";
+import { LocalReferencePosition, PropertySet } from "@fluidframework/merge-tree";
+import { IntervalCollection, SequenceInterval } from "./intervalCollection";
+import { SharedString } from "./sharedString";
+
+const IntervalEventType = {
+	CHANGE: 0,
+	ADD: 1,
+	DELETE: 2,
+	PROPERTYCHANGED: 3,
+} as const;
+
+type IntervalEventType = typeof IntervalEventType[keyof typeof IntervalEventType];
+
+type IntervalRevertible =
+	| {
+			event: typeof IntervalEventType.CHANGE;
+			interval: SequenceInterval;
+			start: LocalReferencePosition;
+			end: LocalReferencePosition;
+			collection: IntervalCollection<SequenceInterval>;
+	  }
+	| {
+			event: typeof IntervalEventType.ADD;
+			interval: SequenceInterval;
+			collection: IntervalCollection<SequenceInterval>;
+	  }
+	| {
+			event: typeof IntervalEventType.DELETE;
+			interval: SequenceInterval;
+			start: LocalReferencePosition;
+			end: LocalReferencePosition;
+			collection: IntervalCollection<SequenceInterval>;
+	  }
+	| {
+			event: typeof IntervalEventType.PROPERTYCHANGED;
+			interval: SequenceInterval;
+			propertyDeltas: PropertySet;
+	  };
+
+type TypedRevertible<T extends IntervalRevertible["event"]> = IntervalRevertible & { event: T };
+
+export function appendLocalAddToRevertibles(
+	interval: SequenceInterval,
+	revertibles: IntervalRevertible[],
+	collection: IntervalCollection<SequenceInterval>,
+) {
+	revertibles.push({
+		event: IntervalEventType.ADD,
+		interval,
+		collection,
+	});
+
+	return revertibles;
+}
+
+export function appendLocalDeleteToRevertibles(
+	string: SharedString,
+	interval: SequenceInterval,
+	revertibles: IntervalRevertible[],
+	collection: IntervalCollection<SequenceInterval>,
+) {
+	// TODO: create localReferences to track positions of deleted interval
+	// just use positions on the interval?
+	revertibles.push({
+		event: IntervalEventType.DELETE,
+		interval,
+		start: interval.start,
```

Comment:
I'm not sure you can just use the existing local references and expect them to still be valid by the time the revert happens. (Also for changeInterval)
_from previous pr_

Reply:
you shouldn't be able to, but I'm glad this got brought up. I think we have a leak here where we're not actually cleaning up local references for deleted/modified intervals. So I wouldn't be surprised if something like this works right now. I'd expect you to need to clone the endpoints here though.

Reply:
https://dev.azure.com/fluidframework/internal/_workitems/edit/4065

Reply:
I couldn't find where they get deleted so I'm also suspicious of a leak. I dunno when a good time to clean those up would be. Maybe when the min seq passes them. Or after the callbacks are invoked, and therefore any listeners should never hang on to them.

This undo-redo code shouldn't be adding to that problem. At the least, any local references created here should get deleted when the revert is invoked. From the app perspective, they probably want to hold on to them for the lifetime of the session. Otherwise maybe we should consider something like a dispose later.

Reply:
in the absence of the undo-redo stack we should just be deleting them immediately on operation, no need to wait for minSeq to pass by.

With undo-redo enabled, that would mean we need to clone refs or create ones at the same positions. Agreed that refs should get deleted when the revert is invoked.